agent/src/beerocks/slave/son_slave_thread.cpp

```diff
@@ -4333,14 +4333,20 @@ bool slave_thread::autoconfig_wsc_parse_m2_encrypted_settings(
 {
     auto encrypted_settings = m2->encrypted_settings();
     uint8_t *iv             = reinterpret_cast<uint8_t *>(encrypted_settings->iv());
-    auto buf                = reinterpret_cast<uint8_t *>(encrypted_settings->encrypted_settings());
+    auto ciphertext         = reinterpret_cast<uint8_t *>(encrypted_settings->encrypted_settings());
+    int clen                = encrypted_settings->encrypted_settings_length();
+    // leave room for up to 16 bytes internal padding length - see aes_decrypt()
+    int dlen = clen + 16;
```

Comment:
That's not true. The decrypted length will always be equal to the encrypted length. The padding is added when you encrypt, not when you decrypt. The decrypted plaintext will include the padding bytes. AFAIK there is no "standard" in AES to detect where the actual plaintext ends and where the padding begins - this is something that has to be encoded in other metadata. In the case of WSC, it's encoded in the attribute length. Anything that comes after the keywrap attribute will be ignored.

Reply:
From https://linux.die.net/man/3/evp_decryptupdate:
> EVP_DecryptInit_ex(), EVP_DecryptUpdate() and EVP_DecryptFinal_ex() are the corresponding decryption operations. EVP_DecryptFinal() will return an error code if padding is enabled and the final block is not correctly formatted. The parameters and restrictions are identical to the encryption operations except that if padding is enabled the decrypted data buffer out passed to EVP_DecryptUpdate() should have sufficient room for (inl + cipher_block_size) bytes unless the cipher block size is 1 in which case inl bytes is sufficient.

According to this, we can either enable padding checks to be done by the library and be on the safe side, like this commit is doing (and therefore the change to the decrypt function). The downside for letting the library do these checks for us is that we need to provide an output buffer which is one block bigger then the input ciphertext, as done here.

However, we can be less strict and just decrypt the data without this check - if we do we can keep the original implementation as it was but just change the prototype to separate the input and output arguments as you suggested in another comment. I'm fine either way - what do you prefer?

Reply:
Coming to think about it again - I think that *this* is the reason Qualcomm always failed to decrypt while the rest of the agents "succeeded" - I believe that Qualcomm uses this capability in the openssl to verify the correctness of the last block.

So - based on that - I think we should be strict when decrypting instead of just ignoring potential wrong padding. It will make it much easier to identify faulty 3rd party agents / controllers.

Reply:
Hm, so apparently I was wrong... I'm pretty sure that specifically for AES, the decrypted size is always equal to the encrypted size, but it's indeed better to follow the EVP_DecryptUpdate constraints.